### PR TITLE
feat: implement walking time to POIs with route calculation and tests

### PR DESCRIPTION
**Pull request by sophieur444**
###  Overview
This PR implements walking time calculation and display for Points of Interest (POIs) near rental listings. Instead of showing straight-line distances, the app now calculates actual walking routes using the OpenRouteService API and displays estimated walking time in minutes (e.g., "10 minutes from EPFL").

This PR corresponds to issue https://github.com/orgs/swent25-team-18/projects/3/views/1?pane=issue&itemId=142126900&issue=swent25-team-18%7Cmy-swiss-dorm%7C210

#### Walking Time Calculation
- **Route-Based Calculation**: Uses OpenRouteService API to calculate actual walking routes between listings and POIs, not just straight-line distances
- **Walking Time Display**: Shows estimated walking time in minutes instead of kilometers in the listing details
- **Smart Caching**: Implements 7-day cache for route calculations to reduce API calls
- **Fallback Mechanism**: Falls back to Haversine distance calculation with route factor if API is unavailable

#### User-Specific University Display
- **Personalized Experience**: If a user has registered a university in their profile, only that university's walking time is shown
- **Default Behavior**: If no university is registered, shows the 2 nearest universities to the listing
- **Seamless Integration**: Automatically fetches user's university from their profile when viewing listings

#### Guaranteed Supermarket Display
- **Always Show Migros**: The nearest Migros supermarket is always displayed, regardless of other POIs
- **Always Show Denner**: The nearest Denner supermarket is always displayed, regardless of other POIs
- **Smart Selection**: Calculates walking time to all Migros/Denner locations and selects the nearest one for each brand

#### POI Grouping
- **Intelligent Grouping**: POIs with the same walking time are automatically grouped together
- **Natural Language**: Displays grouped POIs in a natural format (e.g., "10 minutes from Migros Renens and Denner Renens")
- **Clean UI**: Reduces visual clutter by combining multiple POIs at the same distance

#### Firestore-Based POI Management
- **Universities Collection**: Stores university data (EPFL, UNIL, etc.) in Firestore
- **Supermarkets Collection**: Stores supermarket data (Migros, Denner locations) in Firestore
- **Markets Collection**: Stores market data in Firestore
- **Flexible Data Management**: POI data can be added/updated directly in Firestore without code changes

#### New Services
- **WalkingRouteService**: Core service that interfaces with OpenRouteService API to get walking route distances and convert them to estimated walking times
- **HttpClientProvider**: Centralized OkHttpClient provider for network requests
- **WalkingRouteServiceProvider**: Singleton provider for WalkingRouteService instance

#### New Repositories
- **SupermarketsRepository**: Complete repository pattern implementation for supermarket POIs with Firestore backend
- **MarketsRepository**: Complete repository pattern implementation for market POIs with Firestore backend
- **PointsOfInterestRepository**: Generic POI repository for future extensibility

#### Enhanced Services
- **DistanceService**: 
  - Refactored to use walking time instead of distance
  - Implements complex filtering logic for universities and supermarkets
  - Handles user-specific university preference
  - Ensures guaranteed supermarket display

#### UI Updates
- **ViewListingScreen**: Updated to display walking times with grouped POI formatting
- **ViewListingViewModel**: Enhanced to fetch user profile and pass university preference to distance service
- **String Resources**: Added bilingual support (English/French) for all new POI-related text

###  Testing
- **Unit Tests**: 7 test files covering services, data classes, and business logic
- **Instrumented Tests**: 4 test files covering Firestore repositories and ViewModel integration
- **Edge Cases**: Tests cover invalid coordinates, empty collections, API failures, and user-specific filtering

### API Integration
- **OpenRouteService API**: 
  - Free tier usage (no API key required)
  - Rate limiting: 1 request per second
  - Caching: 7-day cache duration
  - Error handling with graceful fallback

### Data Requirements
- **Firestore Collections**: Requires `universities`, `supermarkets`, and `markets` collections
- **Data Structure**: Each POI must have name, location (with latitude/longitude), and city
- **Security Rules**: Firestore rules must allow read access to POI collections

### Benefits
- **More Accurate Information**: Walking times based on actual routes are more useful than straight-line distances
- **Better User Experience**: Personalized university display and guaranteed supermarket visibility
- **Scalable Architecture**: Repository pattern allows easy extension to new POI types
- **Maintainable Code**: Comprehensive test coverage ensures reliability and easier future modifications
- **Performance Optimized**: Caching and rate limiting reduce API calls and improve response times

Disclaimer: this pr was done with the help/suggestions of ai

Screenshot 2025-12-05 at 03 12 42

### CODE REVIEW

File: app/src/main/java/com/android/mySwissDorm/model/map/WalkingRouteServiceProvider.kt
Comment (karlyounes) [inline app/src/main/java/com/android/mySwissDorm/model/map/WalkingRouteServiceProvider.kt]: This is correct nice work!

File: app/src/main/java/com/android/mySwissDorm/ui/admin/AdminPageScreen.kt [hunk 1]
```diff
@@ -354,6 +354,8 @@ fun AdminPageScreen(
               }
 
               Spacer(Modifier.height(16.dp))
+
+              // POI Data Seeder Section
```
Comment (karlyounes) [inline app/src/main/java/com/android/mySwissDorm/ui/admin/AdminPageScreen.kt]: I don't understand the point of this change

File: app/src/main/java/com/android/mySwissDorm/model/supermarket/SupermarketsRepository.kt
Comment (karlyounes) [inline app/src/main/java/com/android/mySwissDorm/model/supermarket/SupermarketsRepository.kt]: The functions here are good but perhaps we could add a function for exapmle getAllSupermarketsByLocation where you get the supermarkets present in a radius. (you can ignore this comment if you don't find it necessary)

File: app/src/main/java/com/android/mySwissDorm/model/supermarket/SupermarketsRepositoryProvider.kt
Comment (karlyounes) [inline app/src/main/java/com/android/mySwissDorm/model/supermarket/SupermarketsRepositoryProvider.kt]: LGTM

File: app/src/main/java/com/android/mySwissDorm/ui/listing/ViewListingViewModel.kt [hunk 2]
```diff
@@ -166,6 +147,116 @@ class ViewListingViewModel(
...
+              supermarketsRepository = SupermarketsRepositoryProvider.repository,
+              walkingRouteService =
+                  com.android.mySwissDorm.model.map.WalkingRouteServiceProvider.service)
+      val distances = distanceService.calculateDistancesToPOIs(listing.location, userUniversityName)
+      Log.d(
+          "ViewListingViewModel",
+          "Calculated ${distances.size} POI distances for listing ${listing.uid}")
+      Log.d(
+          "ViewListingViewModel",
+          "Listing location: lat=${listing.location.latitude}, lng=${listing.location.longitude}")
```
Comment (karlyounes) [inline app/src/main/java/com/android/mySwissDorm/ui/listing/ViewListingViewModel.kt]: Too many Logs. One is enough

**Review by karlyounes — CHANGES_REQUESTED** [review]
### **Tests:**
The tests look okay with good coverage.

### **Other:**
The POI Repository and repositoryFirestore are not necessary since we already have the Universities and you added the supermarkets so to get the POI close to a residency let's say you could just filter out all the supermarkets and Uni based on location (uni can be fetched from user info). So you can just calculate the distances between supermarkets and the residency from the listing and take the closest one instead of adding a whole repository for it

File: app/src/main/java/com/android/mySwissDorm/ui/admin/AdminPageScreen.kt (hunk 1 above)
Comment (sophieur444) [reply app/src/main/java/com/android/mySwissDorm/ui/admin/AdminPageScreen.kt]: it's a line of a feature that I ended up deleting, I will delete this line thank you!

**Review by karlyounes — CHANGES_REQUESTED** [review]
The changes are good but you still didn't remove the POI type it's not necessary to create a whole type and store it in firestore to get the supermarkets and unis near a residency it's better practice to just use a filter (for supermarkets and universities) to get teh "POI".

**Review by karlyounes — APPROVED** [review]
The changes seem good. Good job not storing the POI in firestore
